src/delay_systems.jl

```diff
@@ -90,93 +95,124 @@ function lsim(sys::DelayLtiSystem{T,S}, u, t::AbstractArray{<:Real}; x0=fill(zer
     else                                            # If u is a regular u(t) function
         (out, t) -> (out .= u(t))
     end
-    _lsim(sys, u!, t, x0, alg; kwargs...)
+
+    _lsim(sys, u!, t, x0, alg; abstol=abstol, reltol=reltol, kwargs...)
 end
 
-function dde_param(dx, x, h!, p, t)
-    A, B1, B2, C2, D21, Tau, u!, uout, hout, tmp = p
+# We actually simulate the integral of y and d here.
+function dde_param(du, u, h, p, t)
+    A, B1, B2, C1, C2, D11, D12, D21, D22, y, Tau, u!, uout, hout, tmpx, tmpy1, tmpy2, tmpd1, tmpnyx, tsave = p
 
-    u!(uout, t)     # uout = u(t)
+    nx = size(A,1)
+    nd = length(Tau)
+    ny = size(C1,1)
+    
+    dx = view(du, 1:nx)
+    dy = view(du, (nx+1):(nx+ny))
+    dd = view(du, (nx+ny+1):(nx+ny+nd))
+    x = view(u, 1:nx)
 
-    #dx .= A*x + B1*ut
+    # uout = u(t)
+    u!(uout, t)
+
+    # hout = d(t-Tau)
+    for k=1:length(Tau)
+        # Get the derivative of the history for d at indices corresponding to Tau
+        hout[k] = h(p, t-Tau[k], Val{1}, idxs=(nx+ny+k))
+    end
+
+    #dx(t) .= A*x(t) + B1*u(t) +B2*d(t-tau)
     mul!(dx, A, x)
-    mul!(tmp, B1, uout)
-    dx .+= tmp
-
-    @views for k=1:length(Tau)     # Add each of the delayed signals
-        u!(uout, t-Tau[k])      # uout = u(t-tau[k])
-        h!(hout, p, t-Tau[k])
-        dk_delayed = dot(C2[k,:], hout) + dot(D21[k,:], uout)
-        dx .+= B2[:, k] .* dk_delayed
+    mul!(tmpx, B1, uout)
+    dx .+= tmpx
+    mul!(tmpx, B2, hout)
+    dx .+= tmpx
```

Comment:
```suggestion
    dx .+= mul!(tmpx, B1, uout)
    dx .+= mul!(tmpx, B2, hout)
```
Perhaps more readable? Or use 5-arg mul, that would save us a lot of tmp variables.

Reply:
I think the 5-arg mul seems nice and fast
```
using LinearAlgebra
using BenchmarkTools

A = randn(10, 10)
B = randn(10, 1)
C = randn(10, 3)

x = randn(10, 1)
u = randn(1, 1)
h = randn(3, 1)

dx = zero(x)
tmp = similar(x)

function f1(A, B, C, x, u, h, tmp)
    mul!(dx, A, x)
    mul!(tmp, B, u)
    dx .+= tmp
    mul!(tmp, C, h)
    dx .+= tmp
end
@benchmark f1($A, $B, $C, $x, $u, $h, $tmp) # ~1 microsecond

function f2(A, B, C, x, u, h, tmp)
    mul!(dx, A, x)
    dx .+= mul!(tmp, B, u)
    dx .+= mul!(tmp, C, h)
end
@benchmark f2($A, $B, $C, $x, $u, $h, $tmp) # ~1 microsecond

function f3(A, B, C, x, u, h)
    mul!(dx, A, x)
    mul!(dx, B, u, 1, 1)
    mul!(dx, C, h, 1, 1)
end
@benchmark f3($A, $B, $C, $x, $u, $h) # ~650 nanoseconds
```